include/pybind11/pytypes.h

```diff
@@ -1285,13 +1312,42 @@ class dict : public object {
 
     size_t size() const { return (size_t) PyDict_Size(m_ptr); }
     bool empty() const { return size() == 0; }
+    detail::dict_accessor operator[](const char *key) const { return {*this, pybind11::str(key)}; }
+    detail::dict_accessor operator[](handle h) const { return {*this, reinterpret_borrow<object>(h)}; }
     detail::dict_iterator begin() const { return {*this, 0}; }
     detail::dict_iterator end() const { return {}; }
     void clear() const { PyDict_Clear(ptr()); }
     template <typename T> bool contains(T &&key) const {
         return PyDict_Contains(m_ptr, detail::object_or_cast(std::forward<T>(key)).ptr()) == 1;
     }
 
+    object get(handle key, handle default_ = none()) const {
+#if PY_MAJOR_VERSION >= 3
+        if (PyObject *result = PyDict_GetItemWithError(m_ptr, key.ptr())) {
+            return reinterpret_borrow<object>(result);
+        } else {
+            if (PyErr_Occurred())
+                throw error_already_set();
+            else
+                return reinterpret_borrow<object>(default_);
+        }
+#else
+        try {
+            return object::operator[](key);
+        } catch (const error_already_set& e) {
+            if (e.type().ptr() == PyExc_KeyError) {
+                return reinterpret_borrow<object>(default_);
+            } else {
+                throw;
+            }
+        }
+#endif
+    }
+
+    object get(const char *key, handle default_ = none()) const {
```

Comment:
To come back to the `setdefault` you mentioned: Python 3 has `PyDict_SetDefault`, but Python 2 doesn't. Since it's also less used than `get`, I'd be fine for leaving this out (for now) and just advising to use `attr("setdefault")(key, default)`.

Reply:
If `PyDict_SetDefault` is not present in Python 2 then I'd also want to leave it out now...

Reply:
Maybe it's again hidden in one way or the other, but ... eh, not worth it. So I agree.

Reply:
> Maybe it's again hidden in one way or the other

Note: there isn't. It's a function with internal linkage, added to the custom methods: https://github.com/python/cpython/blob/2.7/Objects/dictobject.c#L2048 and https://github.com/python/cpython/blob/2.7/Objects/dictobject.c#L2327